File: src/relay/no_relay.py
#!/usr/bin/env python3
"""Tiny TCP relay so an off-box client can reach a localhost-only port.

Nuclear Option binds its remote-command server to 127.0.0.1:5504 (localhost only),
which the vote bot can't reach from another machine. This relay runs INSIDE the
container (launched by the server wrapper) and forwards an externally-reachable
port to that localhost port:

    python3 no_relay.py 0.0.0.0:5550 127.0.0.1:5504

It's stdlib-only and transparent (just pumps bytes both ways), so the bot's
length-prefixed JSON protocol passes straight through.

Keepalive: TCP SO_KEEPALIVE on listen/accept/upstream so idle NAT/LB paths
do not silently drop the sidecar; process still needs an external watchdog if
it exits (ops-watchdog rcmd probe).
"""
import socket
import sys
import threading


def _enable_keepalive(sock):
    try:
        sock.setsockopt(socket.SOL_SOCKET, socket.SO_KEEPALIVE, 1)
    except OSError:
        return
    # Linux: idle 30s, interval 10s, 3 probes. Windows ignores these quietly.
    for opt, val in (
        ("TCP_KEEPIDLE", 30),
        ("TCP_KEEPINTVL", 10),
        ("TCP_KEEPCNT", 3),
    ):
        const = getattr(socket, opt, None)
        if const is None:
            continue
        try:
            sock.setsockopt(socket.IPPROTO_TCP, const, val)
        except OSError:
            pass


def _pump(src, dst):
    try:
        while True:
            data = src.recv(65536)
            if not data:
                break
            dst.sendall(data)
    except OSError:
        pass
    finally:
        for s in (src, dst):
            try:
                s.shutdown(socket.SHUT_RDWR)
            except OSError:
                pass


def _handle(client, target):
    try:
        upstream = socket.create_connection(target, timeout=10)
        _enable_keepalive(upstream)
    except OSError as e:
        sys.stderr.write(f"[relay] upstream connect failed: {e}\n")
        client.close()
        return
    t = threading.Thread(target=_pump, args=(client, upstream), daemon=True)
    t.start()
    _pump(upstream, client)
    for s in (client, upstream):
        try:
            s.close()
        except OSError:
            pass


def main():
    if len(sys.argv) != 3:
        sys.stderr.write("usage: no_relay.py LISTEN_HOST:PORT TARGET_HOST:PORT\n")
        sys.exit(2)
    lh, lp = sys.argv[1].rsplit(":", 1)
    th, tp = sys.argv[2].rsplit(":", 1)
    target = (th, int(tp))

    srv = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    srv.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    _enable_keepalive(srv)
    srv.bind((lh, int(lp)))
    srv.listen(16)
    sys.stdout.write(f"[relay] listening on {lh}:{lp} -> {th}:{tp}\n")
    sys.stdout.flush()
    while True:
        try:
            client, _addr = srv.accept()
            _enable_keepalive(client)
        except OSError as e:
            sys.stderr.write(f"[relay] accept error: {e}\n")
            continue
        threading.Thread(target=_handle, args=(client, target), daemon=True).start()


if __name__ == "__main__":
    main()

File: src/relay/no_relay.pl
#!/usr/bin/perl
# Single-process select() TCP relay (core + Socket only) tuned for Nuclear Option's
# remote-command protocol, where the GAME closes the TCP connection after every
# command. So we keep the client (bot) connection open and open a FRESH upstream to
# the game per request. When the game closes an upstream we DON'T close the client
# (that was dropping the just-written reply through the host's port-forwarder) --
# the client is reused and gets a new upstream on its next request.
#   perl no_relay.pl 0.0.0.0:5550 127.0.0.1:5504
use strict;
use warnings;
use Socket;

my ($lspec, $tspec) = @ARGV;
die "usage: no_relay.pl LHOST:LPORT THOST:TPORT\n" unless $lspec && $tspec;
my ($lh, $lp) = split /:/, $lspec, 2;
my ($th, $tp) = split /:/, $tspec, 2;

sub L { print STDERR "[relay] @_\n"; }

socket(my $srv, PF_INET, SOCK_STREAM, getprotobyname('tcp')) or die "socket: $!";
setsockopt($srv, SOL_SOCKET, SO_REUSEADDR, pack('l', 1));
# TCP keepalive so idle NAT/LB paths do not silently drop the sidecar.
eval { setsockopt($srv, SOL_SOCKET, SO_KEEPALIVE, pack('l', 1)); };
my $laddr = ($lh eq '0.0.0.0' || $lh eq '*' || $lh eq '') ? INADDR_ANY : inet_aton($lh);
bind($srv, sockaddr_in($lp, $laddr)) or die "bind: $!";
listen($srv, 16) or die "listen: $!";
$| = 1;
L("listening on $lspec -> $tspec (per-request upstream, keepalive on)");

my %fh;      # fileno -> filehandle
my %mate;    # fileno -> paired filehandle (cli<->up); a cli may be unpaired (undef)
my %role;    # fileno -> 'cli' | 'up'
my $rin = '';
vec($rin, fileno($srv), 1) = 1;

sub add_fh {
    my ($f, $r) = @_;
    my $n = fileno($f);
    $fh{$n} = $f; $role{$n} = $r; vec($rin, $n, 1) = 1;
    return $n;
}

sub drop_fh {
    my $f = shift;
    return unless defined $f;
    my $n = fileno($f);
    if (defined $n) { vec($rin, $n, 1) = 0; delete $fh{$n}; delete $role{$n}; delete $mate{$n}; }
    close $f;
}

sub open_up {
    my $cli = shift;
    my $up;
    unless (socket($up, PF_INET, SOCK_STREAM, getprotobyname('tcp'))
            && connect($up, sockaddr_in($tp, inet_aton($th)))) {
        L("upstream connect failed: $!");
        return undef;
    }
    eval { setsockopt($up, SOL_SOCKET, SO_KEEPALIVE, pack('l', 1)); };
    my $un = add_fh($up, 'up');
    $mate{fileno($cli)} = $up;
    $mate{$un} = $cli;
    return $up;
}

sub fwd {
    my ($to, $buf, $len) = @_;
    my $off = 0;
    while ($off < $len) {
        my $w = syswrite($to, $buf, $len - $off, $off);
        last if !defined $w;
        $off += $w;
    }
}

while (1) {
    my $rout = $rin;
    my $n = select($rout, undef, undef, undef);
    next if !defined $n || $n <= 0;

    if (vec($rout, fileno($srv), 1)) {
        if (accept(my $cli, $srv)) {
            eval { setsockopt($cli, SOL_SOCKET, SO_KEEPALIVE, pack('l', 1)); };
            add_fh($cli, 'cli');
            $mate{fileno($cli)} = undef;
            L("accepted cli=" . fileno($cli));
        }
    }

    for my $fno (keys %fh) {
        my $f = $fh{$fno};
        next unless defined $f;
        next unless vec($rout, $fno, 1);
        my $who = $role{$fno} || '?';
        my $buf;
        my $r = sysread($f, $buf, 65536);

        if (!defined $r || $r == 0) {                 # closed
            if ($who eq 'up') {                       # game closed this request: keep client
                my $cli = $mate{$fno};
                $mate{fileno($cli)} = undef if defined $cli;
                drop_fh($f);
            } else {                                  # client gone: tear down it + any upstream
                my $up = $mate{$fno};
                drop_fh($f);
                drop_fh($up) if defined $up;
            }
            next;
        }

        if ($who eq 'cli') {
            my $up = $mate{$fno};
            $up = open_up($f) if !defined $up;        # fresh upstream per request
            if (defined $up) { fwd($up, $buf, $r); }
            else { drop_fh($f); }
        } else {                                      # up -> cli
            my $cli = $mate{$fno};
            fwd($cli, $buf, $r) if defined $cli;
        }
    }
}
